drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1322 @@
+	return "";
+}
+
+static bool modem_cellular_gpio_is_enabled(const struct gpio_dt_spec *gpio)
+{
+	return gpio->port != NULL;
+}
+
+static void modem_cellular_enter_state(struct modem_cellular_data *data,
+				       enum modem_cellular_state state);
+
+static void modem_cellular_delegate_event(struct modem_cellular_data *data,
+					  enum modem_cellular_event evt);
+
+static void modem_cellular_event_handler(struct modem_cellular_data *data,
+					 enum modem_cellular_event evt);
+
+static void modem_cellular_dlci1_pipe_handler(struct modem_pipe *pipe,
+					      enum modem_pipe_event event,
+					      void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	switch (event) {
+	case MODEM_PIPE_EVENT_OPENED:
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_DLCI1_OPENED);
+		break;
+
+	default:
+		break;
+	}
+}
+
+static void modem_cellular_dlci2_pipe_handler(struct modem_pipe *pipe,
+					      enum modem_pipe_event event,
+					      void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	switch (event) {
+	case MODEM_PIPE_EVENT_OPENED:
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_DLCI2_OPENED);
+		break;
+
+	default:
+		break;
+	}
+}
+
+static void modem_cellular_chat_callback_handler(struct modem_chat *chat,
+						 enum modem_chat_script_result result,
+						 void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (result == MODEM_CHAT_SCRIPT_RESULT_SUCCESS) {
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS);
+	} else {
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_SCRIPT_FAILED);
+	}
+}
+
+static void modem_cellular_chat_on_imei(struct modem_chat *chat, char **argv, uint16_t argc,
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 2) {
+		return;
+	}
+
+	if (strlen(argv[1]) != 15) {
+		return;
+	}
+
+	for (uint8_t i = 0; i < 15; i++) {
+		data->imei[i] = argv[1][i] - '0';
+	}
+}
+
+static void modem_cellular_chat_on_cgmm(struct modem_chat *chat, char **argv, uint16_t argc,
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 2) {
+		return;
+	}
+
+	strncpy(data->hwinfo, argv[1], sizeof(data->hwinfo) - 1);
+}
+
+static bool modem_cellular_is_registered(struct modem_cellular_data *data)
+{
+	return (data->registration_status_gsm == 1)
+		|| (data->registration_status_gsm == 5)
+		|| (data->registration_status_gprs == 1)
+		|| (data->registration_status_gprs == 5)
+		|| (data->registration_status_lte == 1)
+		|| (data->registration_status_lte == 5);
+}
+
+static void modem_cellular_chat_on_cxreg(struct modem_chat *chat, char **argv, uint16_t argc,
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+	uint8_t registration_status;
+	bool is_registered;
+
```

Comment:
Added the following here (again see my 23-07-27 patch).
```
	if (!data->allow_registration_events) {
		return;
	}
```
During boot we may get network registration events before running our setup scripts. We want to ignore these.
We only care about registration events after we run the `connect_chat_script` script.

Reply:
After the waiting for registration was placed after the dial script, the event is now fired shortly after CFUN=1 is sent. The event is now ignored for all states except AWAIT_REGISTERED and CARRIER_ON, so this addition should not be neccesary hopefully :)